### ISSUE TITLE
.sqlfluffignore rules do not apply when running fix/lint with stdin and --stdin-filename

### ISSUE THREAD

**Issue by peterldowns** [sqlfluff/sqlfluff#6449]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

It seems that by default, [vscode-sqlfluff](https://github.com/sqlfluff/vscode-sqlfluff/) runs `lint` and `fix` commands by passing the file contents over `stdin` and adding the `--stdin-filename <path/to/file/being/edited.sql>` argument. But the filename passed in `--stdin-filename` is not checked against the `.sqlfluffignore` configuration, resulting in fixes/lints being applied to files that have been ignored, just because the contents of the file are passed over stdin.

I'm using sqlfluff-vscode v3.3.0 and sqlfluff 3.2.5.

This results in unexpected behavior when viewing and editing ignored files in vscode. For instance, with a file `schema.sql` that contains lint violations, but whose path is also in the root `.sqlfluffignore` file,

- open `schema.sql` in vscode
  - (correct, expected & observed) see zero lint violations, because vscode-sqlfluff ran `sqlfluff lint --format json <file.sql>`, and the `.sqlfluffignore` rules matched this file path:
- trigger a save in vscode
  - expected: nothing happens, as the file is ignored and lint/fix rules should not be applied
  - observed: the file is linted with `sqlfluff fix --stdin-filename <file.sql> -`, and the `.sqlfluffignore` rules are not applied to the stdin-filename, so the file ends up overwritten by the extension.

---

Before [--stdin-filename was added](https://github.com/sqlfluff/sqlfluff/pull/5805), the maintainers [responded to this issue](https://github.com/sqlfluff/sqlfluff/issues/1038) and wrote:

> When using stdin, SQLFluff oughtn't have a clue where the source string came from, so it should lint the data even if that data happens to be from an ignored file. Practically, stdin data does not have a filename, so there would be no way to check if it is ignored.

Now that `--stdin-filename` exists, would you be open to a PR that checks them against the root `.sqlfluffignore` file? Or, does the full directory walk like `paths_from_path` and picks up on nested config/ignore files to make sure they're applied appropriately?

### Expected Behaviour

The expected behavior is to ignore the file contents being passed in because they are from a file that has been ignored. `sqlfluff fix` should return the contents unchanged. `sqlfluff lint` should perform 0 linting and return successfully without any errors.

```shell
# cat pkg/database/schema.sql | sqlfluff fix --stdin-filename pkg/database/schema.sql -
SELECT col_a a FROM foo; -- lint error, missing AS

# cat pkg/database/schema.sql | sqlfluff lint --stdin-filename pkg/database/schema.sql -
All Finished 📜 🎉!
```

### Observed Behaviour

```shell
# cat pkg/database/schema.sql | sqlfluff fix --stdin-filename pkg/database/schema.sql -
SELECT col_a AS a FROM foo; -- lint error, missing AS

# cat pkg/database/schema.sql | sqlfluff lint --stdin-filename pkg/database/schema.sql -
== [stdin] FAIL
L:   1 | P:  14 | AL02 | Implicit/explicit aliasing of columns.
                       | [aliasing.column]
All Finished 📜 🎉!
```

### How to reproduce

You can clone this repository and run the commands described in its readme:

https://github.com/peterldowns/repro-sqlfluff-ignore-stdin-filename

```shell
# cat pkg/database/schema.sql | sqlfluff fix --stdin-filename pkg/database/schema.sql -
SELECT col_a AS a FROM foo; -- lint error, missing AS

# cat pkg/database/schema.sql | sqlfluff lint --stdin-filename pkg/database/schema.sql -
== [stdin] FAIL
L:   1 | P:  14 | AL02 | Implicit/explicit aliasing of columns.
                       | [aliasing.column]
All Finished 📜 🎉!
```

### Dialect

postgres

### Version

```shell
# python3 --version
Python 3.11.0

# sqlfluff --version
sqlfluff, version 3.2.5
```

### Configuration

```shell
# cat .sqlfluff
[sqlfluff]
dialect = postgres

# cat .sqlfluffignore
pkg/database/schema.sql
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterldowns** [comment]
Bump — I'm willing to work on a PR if you're interested in having this fixed. Can a maintainer please let me know?

**Comment by keraion** [comment]
I agree with your reasoning. With the passed in file path, we should be able to honor the .sqlfluffignore file.

**Comment by peterldowns** [comment]
Thank you @keraion and @WittierDinosaur!
